File: lib-java/src/collections/ranged/StaticPrefixSum.java
package collections.ranged;

class StaticPrefixSum {
  private long[] prefix;

  StaticPrefixSum(int[] array) {
    prefix = new long[array.length + 1];
    for (int i = 1; i <= array.length; i++) {
      prefix[i] = prefix[i - 1] + array[i - 1];
    }
  }

  StaticPrefixSum(long[] array) {
    prefix = new long[array.length + 1];
    for (int i = 1; i <= array.length; i++) {
      prefix[i] = prefix[i - 1] + array[i - 1];
    }
  }

  long rangeSum(int a, int b) {
    return prefix[b] - prefix[a];
  }
}

File: Library/cs/Collections/Ranged/StaticPrefixSum.cs
using System.Numerics;

namespace Evermore.Collections.Ranged;

public class StaticPrefixSum<T> where T : IBinaryInteger<T>
{
    private long[] _prefix;
    public StaticPrefixSum(T[] array)
    {
        _prefix = new long[array.Length + 1];

        for (int i = 1; i <= array.Length; i++)
        {
            long currentVal = long.CreateChecked(array[i - 1]);
            _prefix[i] = _prefix[i - 1] + currentVal;
        }
    }

    public long RangeSum(int a, int b)
    {
        return _prefix[b] - _prefix[a];
    }
}
